### PR TITLE
refactor: modernize engine with Pydantic models, modular architecture, and session recovery

### PR DESCRIPTION
**Pull request by anticomputer**
The engine code was a single 671-line `__main__.py` with raw dict access everywhere, no type safety, and no error recovery. This rewrites the internals while keeping the YAML grammar fully backwards-compatible. All existing taskflows (both example and production seclab-taskflows) work without changes.

## What changed

**Architecture.** `__main__.py` is now 25 lines. The logic lives in focused modules:
- `models.py` -- Pydantic v2 models for all five YAML document types (taskflow, personality, toolbox, model_config, prompt). Validation happens at parse time instead of scattering `.get()` calls everywhere.
- `runner.py` -- execution engine, extracted from `__main__` and decomposed into `_resolve_model_config()`, `_merge_reusable_task()`, `_resolve_task_model()`, `_build_prompts_to_run()`.
- `cli.py` -- Typer-based CLI replacing argparse. Adds `--debug` and `--resume` flags.
- `agent.py` -- TaskAgent wrapper, now supports per-model API type, endpoint, and token overrides.
- `mcp_lifecycle.py`, `mcp_transport.py`, `mcp_prompt.py` -- decomposed from the old 462-line `mcp_utils.py`.
- `session.py` -- task-level checkpoint/resume with auto-retry.
- `prompt_parser.py` -- extracted to break a circular import between cli and runner.

**Pydantic grammar models.** Every YAML file is validated against a typed model at load time. `extra="allow"` on all models so new fields do not break old parsers. Reserved word conflicts handled via aliases (`async` -> `async_task`, `model_config` -> `model_config_ref`). Version normalization (`1` -> `"1.0"`) with validation.

**Per-model API configuration.** `model_settings` in a model_config can now include `api_type` (chat_completions or responses), `endpoint`, and `token` (env var name). These are extracted by the engine before passing remaining settings to the SDK. This lets you mix endpoints and API types across models in a single taskflow.

**Arbitrary API endpoints.** Unknown API endpoints no longer crash the agent at import time or when listing models. The model catalog and tool-call detection fall back gracefully for non-standard endpoints, so users can point at any compatible API.

**Session recovery.** Taskflow runs are checkpointed after each task. On unrecoverable failure, the session ID is printed and `--resume` picks up from the last successful task. Failed `must_complete` tasks are not advanced past on resume. Failed tasks auto-retry 3 times with backoff before giving up.

**Error output.** Exceptions now print a concise one-line chain by default instead of dumping 80-line tracebacks through httpx/httpcore/openai internals. `--debug` or `TASK_AGENT_DEBUG=1` gets you the full trace. Setting `TASK_AGENT_DEBUG=0` or `TASK_AGENT_DEBUG=false` correctly disables debug mode.

**Code quality.** Type hints and docstrings on all public functions. `__all__` exports on every module. Fixed mutable default args in `TaskAgent.__init__`. Replaced bare `except Exception` with specific types. Zero ruff errors across the entire codebase. Ruff ignores reduced from 59 to 21.

## What did not change

The YAML grammar is 100% backwards-compatible. No taskflow files need modification. All field names, nesting, template syntax, and behavior are preserved.

## Testing

Unit tests for Pydantic models, runner internals (model resolution, task merging, prompt building, pop-after-render correctness), Typer CLI arg parsing, debug env handling, prompt parser edge cases, capi endpoint handling with arbitrary endpoint fallback, and session persistence including corrupt file recovery. 155 tests total, all passing, ruff clean.

Integration tested against api.githubcopilot.com:

- All example taskflows pass (echo, globals, reusable prompts, reusable tasks, repeat_prompt, responses API)
- Comprehensive test taskflow exercising every feature: shell tasks, repeat_prompt, async iteration, model_config, globals, inputs, env, MCP toolboxes, headless, blocked_tools, reusable tasks, reusable prompts, and agent handoffs
- Edge case test taskflow: nested JSON repeat_prompt, sequential iteration, task-level env scoping, globals CLI override, max_steps
- Production seclab-taskflows tested: fetch_source_code, classify_application, classify_application_local, fetch_audit_issue

### CODE REVIEW

File: src/seclab_taskflow_agent/__main__.py [hunk 1]
```diff
@@ -1,671 +1,27 @@
...
-                task_complete = await run_prompts(async_task=async_task, max_concurrent_tasks=max_concurrent_tasks)
-
-                if must_complete and not task_complete:
-                    logging.critical("Required task not completed ... aborting!")
-                    await render_model_output("🤖💥 *Required task not completed ...\n")
-                    break
-
+# Re-export for backwards compatibility — some tests import from __main__
+from .prompt_parser import parse_prompt_args  # noqa: E402, F401
+from .runner import deploy_task_agents, run_main  # noqa: E402, F401
```
[the remark this answers is not in the document]
Comment (anticomputer) [reply src/seclab_taskflow_agent/__main__.py]: Intentional re-exports for backwards compatibility. Uses explicit `as X` syntax recognized by ruff/pyflakes.

File: src/seclab_taskflow_agent/mcp_transport.py [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+    async def call_tool(self, *args: Any, **kwargs: Any) -> Any:
+        """Call a tool via the dedicated loop (serialised with a lock)."""
+        async with self.lock:
+            return asyncio.run_coroutine_threadsafe(super().call_tool(*args, **kwargs), self.t.loop).result()
+
+    async def cleanup(self, *args: Any, **kwargs: Any) -> None:
+        """Clean up and shut down the dedicated loop."""
+        try:
+            asyncio.run_coroutine_threadsafe(super().cleanup(*args, **kwargs), self.t.loop).result()
+        except asyncio.CancelledError:
```
[the remark this answers is not in the document]
Comment (anticomputer) [reply src/seclab_taskflow_agent/mcp_transport.py]: Intentional — this swallows `CancelledError` during the async event loop shutdown in `AsyncDebugMCPServerStdio.cleanup()`. The comment "Swallow cancellation during cleanup shutdown" is on the same line.

File: src/seclab_taskflow_agent/capi.py [hunk 3]
```diff
@@ -73,10 +79,8 @@ def list_capi_models(token: str) -> dict[str, dict]:
...
                 models_catalog = "models"
             case _:
-                raise ValueError(
-                    f"Unsupported Model Endpoint: {api_endpoint}\n"
-                    f"Supported endpoints: {[e.to_url() for e in AI_API_ENDPOINT_ENUM]}"
-                )
+                # Unknown endpoint — try the OpenAI-style models catalog
+                models_catalog = "models"
         r = httpx.get(
             httpx.URL(api_endpoint).join(models_catalog),
```
[the remark this answers is not in the document]
Comment (anticomputer) [reply src/seclab_taskflow_agent/capi.py]: Pre-existing behavior -- the httpx URL join pattern is unchanged from the capi.py on main. The known endpoints (CAPI, GitHub Models) do not include path segments so this works in practice, but it is worth hardening for arbitrary endpoints in a follow-up.

File: src/seclab_taskflow_agent/render_utils.py [hunk 4]
```diff
@@ -16,17 +20,19 @@
...
-async def flush_async_output(task_id: str):
+async def flush_async_output(task_id: str) -> None:
+    """Flush buffered async output for *task_id* to the console."""
     async with async_output_lock:
         if task_id not in async_output:
-            raise ValueError(f"No async output for task: {task_id}")
-        data = async_output[task_id]
-        del async_output[task_id]
+            # No buffered output (agent may have failed before producing any).
+            return
```
[the remark this answers is not in the document]
Comment (anticomputer) [reply src/seclab_taskflow_agent/render_utils.py]: The old code raised `ValueError` if the task_id wasn't in the buffer. But an agent can fail before producing any output (e.g. connection error during setup), and `flush_async_output` is called unconditionally in the cleanup path. Raising there would mask the original failure. The early return with the comment on line 27 makes this a graceful no-op instead.

File: src/seclab_taskflow_agent/mcp_transport.py [hunk 5]
```diff
@@ -0,0 +1,313 @@
...
+        poll_interval: float = 0.5,
+        env: dict[str, str] | None = None,
+    ) -> None:
+        super().__init__(daemon=True)
+        self.url: str = url
+        self.cmd: list[str] = cmd
+        self.on_output: Callable[[str], None] | None = on_output
+        self.on_error: Callable[[str], None] | None = on_error
+        self.poll_interval: float = poll_interval
+        self.env: dict[str, str] = _filtered_env()
```
[the remark this answers is not in the document]
Comment (anticomputer) [reply src/seclab_taskflow_agent/mcp_transport.py]: MCP subprocesses need a working base environment (`PATH`, `HOME`, `SHELL`, `LANG`, etc.) to function — without these, even basic commands and process startup fail. The toolbox YAML `env:` section is for adding server-specific variables on top of this inherited base. This matches the previous behavior (the old code had `self.env = os.environ.copy()` with a `# XXX: potential for environment leak` comment) but improves on it with the `TASKFLOW_ENV_DENYLIST` mechanism for filtering sensitive variables. Switching to an allowlist-only approach would break existing toolbox configurations that rely on a working shell environment.
